esphome/components/image/__init__.py

```diff
@@ -50,86 +103,161 @@
 async def to_code(config):
     from PIL import Image
 
-    path = CORE.relative_config_path(config[CONF_FILE])
-    try:
-        image = Image.open(path)
-    except Exception as e:
-        raise core.EsphomeError(f"Could not load image file {path}: {e}")
+    if CONF_FILE in config:
+        path = CORE.relative_config_path(config[CONF_FILE])
+        try:
+            image = Image.open(path)
+        except Exception as e:
+            raise core.EsphomeError(f"Could not load image file {path}: {e}")
+    elif CONF_MDI in config:
+        mdi_id = config[CONF_MDI]
```

Comment:
Was already loving this PR, but seeing this get added just made my day. I recently started working on an `icon_font` component so I could easily use MDI icons, but I like this better. 

One design decision I made early on that I'd like you to consider is supporting using an optional prefix in the icon name - ie, allowing both `weather-sunny` and `mdi:weather-sunny`. So the YAML could be written as:

```yaml
image:
  - id: mdi_weather_sunny
    mdi: "mdi:weather-sunny"
```

Aside from being the same syntax that Home Assistant uses, it allows using existing editor support for autocomplete, preview, and linting  - see [this extension](https://marketplace.visualstudio.com/items?itemName=lukas-tr.materialdesignicons-intellisense) for support for this in VS Code. Not everyone will use that, so I think the prefix should be optional.

Conviently, ESPHome requires Python 3.9 (as of 2022.10.0), which added `removeprefix()` (it just returns the original string if the prefix isn't there)

```suggestion
        mdi_id = config[CONF_MDI].removeprefix("mdi:")
```

Reply:
The MDI support is actually a separate PR (#4654); this one (online images) just happens to build on top of that one.
I haven't played with the VS Code extension yet, but from the documentation, it looks like it should be possible to enhance the regex used to detect mdi; why not add a "mdi: {param}" (i.e. with the space) to the extension configuration instead? would make the yaml better readable (`mdi: mdi:whatever` looks very ugly IMHO).

Reply:
Ah, I missed that, thanks.

I know it's not the prettist, but it's a simple win. 

TBH, if I were the maintainer of that extension, I would suggest ESPHome adopt the existing standard, rather than add yet another variation to the extension and risk breaking something unexpected.

Reply:
Well, it is a separate PR, but this one (#4710) is rebased on top of that one, so yes, it can be tested by pointing it at this one ;-)

And I am also the creator of the other PR, so you still reached the "right" guy :-D

I'm still not fully convinced it looks good on the yaml config, but haven't been able to make the one with the space work with intellisense, so I'll add that to the PR, and let the reviewers decide whether to keep it in or not. As you said, it doesn't really hurt; nobody forces you to use the ugly "mdi: mdi:icon" syntax.